### PR TITLE
Layernorm: recognize explicit pattern and transform it to a layer norm operation (recompose)

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
Recognize Roberta's expanded layer norm and recompose it into a layer norm.

```
    %mean = "onnx.ReduceMeanV13"(%x) {axes = [-1], keepdims = 1 : si64} : (tensor<1x384x768xf32>) -> tensor<1x384x1xf32>
    %d = "onnx.Sub"(%x, %mean) : (tensor<1x384x768xf32>, tensor<1x384x1xf32>) -> tensor<1x384x768xf32>
    %dd = "onnx.Mul"(%d, %d) : (tensor<1x384x768xf32>, tensor<1x384x768xf32>) -> tensor<1x384x768xf32>
    %var = "onnx.ReduceMeanV13"(%dd) {axes = [-1], keepdims = 1 : si64, onnx_node_name = "ReduceMean_42"} : (tensor<1x384x768xf32>) -> tensor<1x384x1xf32>
    %varEps = "onnx.Add"(%var, %eps) : (tensor<1x384x1xf32>, tensor<f32>) -> tensor<1x384x1xf32>
    %StdDev = "onnx.Sqrt"(%varEps) : (tensor<1x384x1xf32>) -> tensor<1x384x1xf32>
    %Norm = "onnx.Div"(%d, %StdDev) : (tensor<1x384x768xf32>, tensor<1x384x1xf32>) -> tensor<1x384x768xf32>
    %NormScaled = "onnx.Mul"(%Norm, %scale) : (tensor<1x384x768xf32>, tensor<768xf32>) -> tensor<1x384x768xf32>
    %Y = "onnx.Add"(%NormScaled, %bias) : (tensor<1x384x768xf32>, tensor<768xf32>) -> tensor<1x384x768xf32>
```

becomes
```
  %Y, %Mean, %InvStdDev = "onnx.LayerNormalization"(%0, %arg1, %arg2) {axis = 2 : si64, epsilon = 9.99999974E-6 : f32, stash_type = 1 : si64} : (tensor<1x384x768xf32>, tensor<768xf32>, tensor<768xf32>) -> (tensor<1x384x768xf32>, none, none)

```

Added a new optimization phase, `recompose` that is run separately from decompose, run just after it.

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
Results in 5% faster exec time on z16 with NNPA.

With LayerNorm (executed on CPU with optimized code)
```
  zhigh.Unstick, 109, 0.0005423, 0.0591120, 26.1%
  zhigh.MatMul, 97, 0.0004878, 0.0473119, 20.9%
  onnx.Erf, 12, 0.0038903, 0.0466839, 20.6%
  zhigh.Stick, 109, 0.0001479, 0.0161256, 7.1%
  zhigh.Add, 108, 0.0001403, 0.0151509, 6.7%
  onnx.Transpose, 48, 0.0002647, 0.0127078, 5.6%
  zhigh.Softmax, 12, 0.0010360, 0.0124321, 5.5%
  zhigh.Div, 24, 0.0001799, 0.0043186, 1.9%
  onnx.LayerNormalization, 25, 0.0001272, 0.0031794, 1.4%
  zhigh.Mul, 24, 0.0001222, 0.0029339, 1.3%
...
Statistics end, all ops, ordered_by time, tot_time 0.2266896
```

Without Layernorm (code corresponding to layer norm executed on NNPA and CPU)
```
Statistics start, all ops, ordered_by time, tot_time 0.2390143
  zhigh.Unstick, 110, 0.0005370, 0.0590718, 24.7%
  onnx.Erf, 12, 0.0038874, 0.0466486, 19.5%
  zhigh.MatMul, 97, 0.0003957, 0.0383823, 16.1%
  zhigh.Add, 133, 0.0001406, 0.0186982, 7.8%
  zhigh.Stick, 110, 0.0001465, 0.0161152, 6.7%
  onnx.Transpose, 48, 0.0003163, 0.0151814, 6.4%
  zhigh.Softmax, 12, 0.0010363, 0.0124359, 5.2%
  onnx.Div, 25, 0.0004263, 0.0106571, 4.5%
  onnx.Sub, 25, 0.0002008, 0.0050194, 2.1%
  zhigh.Div, 24, 0.0001785, 0.0042848, 1.8%
  zhigh.Mul, 49, 0.0000781, 0.0038284, 1.6%
```

File: src/Transform/ONNX/Recompose.cpp [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+  Value dummy;
+  if (operandOfOpDefinedBy<OP>(op, dummy, matchOp, matchThisOperandIndex)) {
+    matchOperand0 = op->getOperand(0);
+    matchOperand1 = op->getOperand(1);
+    return true;
+  }
+  return false;
+}
+
+struct RecomposeLayerNormFromAddPattern : public OpRewritePattern<ONNXAddOp> {
```
Comment (tungld) [inline src/Transform/ONNX/Recompose.cpp:95]: Do you think this rewrite patten can be moved to Rewrite.cpp? It's like fusing Add into LayerNorm.

**Review by tungld — APPROVED** [review]
LGTM!

File: src/Transform/ONNX/Recompose.cpp (hunk 1 above)
Comment (AlexandreEichenberger) [reply src/Transform/ONNX/Recompose.cpp:95]: good suggestion.

Comment (AlexandreEichenberger) [reply src/Transform/ONNX/Recompose.cpp]: done

**Pull request merged**